File: 1189_max_num_of_balloons/max_number_of_balloons.py
from collections import Counter


def max_number_of_balloons(text: str) -> int:
    """
    Time: O(n)
    Space: O(n)
    """
    str_count = Counter(text)
    balloon_count = Counter('balloon')
    return min(str_count[c] // balloon_count[c] for c in balloon_count)
       

File: 1189_max_num_of_balloons/max_number_of_balloons.java
class Solution {
    // Time: O(n)
    // Space: O(1)
    public int maxNumberOfBalloons(String text) {
        int[] strCount = count(text);
        int[] balloonCount = count("balloon");
        int res = text.length();
        for (char c : "balloon".toCharArray()) {
            res = Math.min(res, strCount[c - 'a'] / balloonCount[c - 'a']);
        }
        return res;
    }
    
    private int[] count(String str) {
        int[] res = new int[26];
        for (int i = 0; i < str.length(); i++) {
            res[str.charAt(i) - 'a']++;
        }
        return res;
    }
}